### PR TITLE
Add caching helper methods, and caching to API calls

### PR DESCRIPTION
**Pull request by samwilson**
This creates a few methods in a new Helper parent class, to make
it slightly easier to add caching to any helper.
Also some documentation.

Bug: https://phabricator.wikimedia.org/T161057

The specific TTLs here can be changed of course; I'm not sure what they should be. It depends on how people use the tools, but at least some things can be cached for quite a long while (to speed things up).

### CODE REVIEW

File: src/AppBundle/Helper/ApiHelper.php [hunk 1]
```diff
@@ -50,6 +51,7 @@ public function groups($project, $username)
             if (isset($res["batchcomplete"]) && isset($res["query"]["users"][0]["groups"])) {
                 $result = $res["query"]["users"][0]["groups"];
             }
+            $this->cacheSave($cacheKey, $result, 'P7D');
```
Comment (MusikAnimal) [inline src/AppBundle/Helper/ApiHelper.php]: This info along with global user groups indeed rarely change, but unlike the namespaces this same info may not be requested over and over. Given how popular the tool is, is it worth caching that info in Redis for that long?

File: src/AppBundle/Helper/EditCounterHelper.php [hunk 2]
```diff
@@ -143,6 +140,7 @@ public function getRevisionCounts($userId)
             3
         );
 
+        $this->cacheSave($cacheKey, $revisionCounts, 'P1D');
```
Comment (MusikAnimal) [inline src/AppBundle/Helper/EditCounterHelper.php]: This I don't think should be cached at all. Revision counts change frequently and people will probably complain that the numbers are off

**Comment by MusikAnimal** [comment]
Is it possible to cache some things only in development (since I know that would help us)? People will want the live data in production. Frankly I'd say namespaces are the only thing worth caching. User groups and global groups don't change much, but I'm willing to be most of the time the cache will go unused, and they're not especially slow queries.

File: src/AppBundle/Helper/ApiHelper.php (hunk 1 above)
Comment (samwilson) [inline src/AppBundle/Helper/ApiHelper.php]: I figured anything that reduces the number of API calls is good, but I'm happy either way. Will remove; good to keep it simple anyway. :)

File: src/AppBundle/Helper/EditCounterHelper.php (hunk 2 above)
Comment (samwilson) [inline src/AppBundle/Helper/EditCounterHelper.php]: Yes, good point.

**Comment by samwilson** [comment]
>Is it possible to cache some things only in development (since I know that would help us)? 

Not sure, but it's a good idea. I think we'd have to specify a different cache adapter wherever we want to do dev-only caching.

I'll pare it back to the minimum for now.

File: src/AppBundle/Helper/HelperBase.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+     */
+    protected function cacheHas($key)
+    {
+        /** @var \Symfony\Component\Cache\Adapter\AdapterInterface $cache */
+        $cache = $this->container->get('cache.app');
+        return $cache->getItem($this->getCacheKey($key))->isHit();
+    }
+
+    /**
+     * Get a value from the cache. This is
```
Comment (kaldari) [inline src/AppBundle/Helper/HelperBase.php]: "This is"?

File: docs/development.rst [hunk 4]
```diff
@@ -46,7 +44,37 @@ And set the following in ``app/config/parameters.yml``:
...
+=======
+
+Caching should happen in helpers, with appropriate times-to-live.
+
+Every helper should extend HelperBase, which has ``cacheHas()``, ``cacheGet()``, and ``cacheSave()`` methods.
+These should be used in this pattern::
+
+    public function doSomething($input)
+    {
+        $cacheKey = 'ec.something.'.$input;
```
Comment (kaldari) [inline docs/development.rst]: What's "ec"? Just curious :)

**Pull request merged**